homeassistant/components/notify/__init__.py

```diff
@@ -4,110 +4,281 @@
 For more details about this component, please refer to the documentation at
 https://home-assistant.io/components/notify/
 """
-from functools import partial
-import logging
+import io
 import os
+import logging
 
 import voluptuous as vol
+import requests
 
-import homeassistant.bootstrap as bootstrap
 from homeassistant.config import load_yaml_config_file
-from homeassistant.helpers import config_per_platform, template
+from homeassistant.helpers import template
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
 from homeassistant.helpers.config_validation import PLATFORM_SCHEMA  # noqa
 import homeassistant.helpers.config_validation as cv
-from homeassistant.const import CONF_NAME
+from homeassistant.const import (ATTR_LATITUDE, ATTR_LONGITUDE, ATTR_ENTITY_ID,
+                                 STATE_UNKNOWN)
 
-DOMAIN = "notify"
+DOMAIN = 'notify'
 
 # Title of notification
-ATTR_TITLE = "title"
+ATTR_TITLE = 'title'
 ATTR_TITLE_DEFAULT = "Home Assistant"
 
 # Target of the notification (user, device, etc)
 ATTR_TARGET = 'target'
 
 # Text to notify user of
-ATTR_MESSAGE = "message"
+ATTR_MESSAGE = 'message'
 
 # Platform specific data
-ATTR_DATA = 'data'
+ATTR_FORMATS = 'formats'
+
+ATTR_PHOTO = 'photo'
+
+# others
+ATTR_FILE = 'file'
+ATTR_URL = 'url'
+ATTR_CAPTION = 'caption'
+ATTR_USERNAME = 'username'
+ATTR_PASSWORD = 'password'
+
+# HASS Device objects
+ATTR_CAMERA = 'camera_entity'
+ATTR_DEVICE = 'device_entity'
 
-SERVICE_NOTIFY = "notify"
+SERVICE_SEND_MESSAGE = 'send_message'
+SERVICE_SEND_PHOTO = 'send_photo'
+SERVICE_SEND_LOCATION = 'send_location'
 
-NOTIFY_SERVICE_SCHEMA = vol.Schema({
+MESSAGE_SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
     vol.Required(ATTR_MESSAGE): cv.template,
-    vol.Optional(ATTR_TITLE, default=ATTR_TITLE_DEFAULT): cv.string,
+    vol.Optional(ATTR_TITLE, default=ATTR_TITLE_DEFAULT): cv.template,
     vol.Optional(ATTR_TARGET): cv.string,
-    vol.Optional(ATTR_DATA): dict,
```

Comment:
I think we still want to keep this as it will allow platforms to leverage special things that are not perse standardized.